README.md
# immutable-core-model

Immutable Core Model integrates with the
[Immutable App](https://www.npmjs.com/package/immutable-app) ecosystem and
provides persistence of immutable data objects using MySQL, Redis and
Opensearch.

Immutable Core Model requires native async/await support.

## Creating a new database connection

    const mysql = ImmutableCoreModel.createMysqlConnection({
        database: 'database-name',
        host: 'localhost',
        password: 'db-password'
        user: 'db-user'
    })

The connection parameters will have required defaults added and will then
be passed to mysql2 to create a connection.

If the `connectionLimit` param is set then a connection pool will be created
instead of a single connection. This is recommended for production use.

The default connection params are:

    bigNumberStrings: true,
    dateStrings: true,
    namedPlaceholders: true,
    rowsAsArray: false,
    supportBigNumbers: true,

These parameters are needed for Immutable Core Model process results correctly
and so they cannot be changed.

## Creating a new model

    const ImmutableCoreModel = require('immutable-core-model')

    var fooModel = new ImmutableCoreModel({name: 'foo'})

## Syncing model specification to database

    await fooModel.sync()

In a production environment clients should not have database credentials that
allow anything besides INSERT and SELECT on the database.

All models should be sync'd prior to deploying new code with specification
changes that require creating or modifying tables.

## Altering model tables, columns, and indexes

Immutable Core Model only supports limited alterations to existing model
tables:

* Adding new columns
* Adding new indexes to columns without indexes

These operations are currently done as ALTER TABLE statements which may cause
significant performance impacts with large tables in production envrionments.

## Compression

By default Immutable Core Model compresses data records using Google's snappy compression algorithm.

### Disabling compression for an individual model

    new ImmutableCoreModel({compression: false})

### Setting the default compression setting for all models

    ImmutableCoreModel.defaultCompression(false)

### Setting compression from the env

    DEFAULT_COMPRESSION=true node app
    DEFAULT_COMPRESSION=false node app

    DEFAULT_COMPRESSION=1 node app
    DEFAULT_COMPRESSION=0 node app

If the default compression setting is set in the environment this will
override any values set either globally or at the model level in code.

## Id column type

By default Immutable Core Model uses binary columns for ids. This saves space
but makes it more difficult to work with the database using other tools.

### Disabling binaryIds for an individual model

    new ImmutableCoreModel({binaryIds: false})

### Setting the default binaryIds setting for all models

    ImmutableCoreModel.defaultBinaryIds(false)

### Setting id column type from the env

    DEFAULT_BINARY_IDS=true node app
    DEFAULT_BINARY_IDS=false node app

    DEFAULT_BINARY_IDS=1 node app
    DEFAULT_BINARY_IDS=0 node app

If the default binary ids setting is set in the environment this will
override any values set either globally or at the model level in code.

## Schema

For the simple example above the following schema will be created:

    CREATE TABLE foo (
        n bigint(20) unsigned NOT NULL AUTO_INCREMENT,
        c smallint(5) unsigned NOT NULL DEFAULT '1',
        d tinyint(1) NOT NULL DEFAULT '0',
        fooAccountId binary(16) NOT NULL,
        fooCreateTime datetime(6) NOT NULL,
        fooData mediumblob NOT NULL,
        fooId binary(16) NOT NULL,
        fooOriginalId binary(16) NOT NULL,
        fooParentId binary(16) DEFAULT NULL,
        fooSessionId binary(16) NOT NULL,
        PRIMARY KEY (n),
        UNIQUE KEY (fooId),
        UNIQUE KEY (fooParentId),
        KEY (fooAccountId),
        KEY (fooCreateTime),
        KEY (fooOriginalId),
        KEY (fooSessionId)
    ) ENGINE=InnoDB DEFAULT CHARSET=utf8

All single character column names are reserved for system use and single
character system columns will always come first in the schema.

All model columns are created in alphabetical order after system columns.

As of version 2.0.0 all ids are lower case. Previously they were upper case.

### n

The n column provides and auto increment id for each record.

**NEVER USE THIS ID IN YOUR APPLICATIONS**

### c

The c column indicates whether or not record data is compressed. The c
column will only be created if a data column is present.

0 indicates that data is not commpressed and 1 indicates that data is
compressed using snappy.

### d

The d column indicates whether or not the record is deleted.

0 indicates that the record is not deleted and 1 indicates that the record
is deleted.

### fooAccountId

The accountId of the account that owns the foo object.

### fooCreateTime

fooCreateTime is a microsecond timestamp generated using the micro-timestamp
module.

Timestamps are generated in the application and not in the database which is
necessary because fooId includes the timestamp in the data that is hashed.

### fooData

fooData is a JSON encoding of the foo object using the
[json-stable-stringify](https://www.npmjs.com/package/json-stable-stringify)
module. This data is then compressed using the
[snappy](https://www.npmjs.com/package/snappy) module which is a binding for
Google's Snappy compression library.

### fooId

fooId is a hash of the JSON encoding of fooOriginalId, fooParentId, fooData,
fooCreateTime, sessionId, and accountId. The hash is calculated before any
compression is performed.

Undefined values are not included in the JSON encoding and the database driver
converts NULL values from the database to undefined so that they will not be
included in JSON encodings.

The hash value is the first 128 bits on an SHA-2 hash calculated using the
[stable-id](https://www.npmjs.com/package/stable-id) module.

### fooOriginalId and fooParentId

The original and parent ids in Immutable Core Model are used to track object
revisions.

When a foo record is initially created its originalId will be equal to its id
and its parentId will be null.

When a revision to that record is created it will have the same originalId as
the first record and its parentId will be equal to the id of the first record.

#### fooId, fooOriginalId, fooParentId Example

Revision | fooId | fooOriginalId | fooParentId |
---------|-------|---------------|-------------|
1st      | 1111  | 1111          | NULL        |
2nd      | 2222  | 1111          | 1111        |
3rd      | 3333  | 1111          | 2222        |
4th      | 4444  | 1111          | 3333        |

#### Distributed Concurrency Control

The UNIQUE INDEX on parentFooId insures that in a multi-reader, multi-writer,
distributed system data does not become corrupted.

In order to create a new revision of foo it is necessary to provide the
parentFooId from the last version of foo fetched from the database.

If this record is outdated then the INSERT with that parentFooId will fail
because it has already been used by another writer.

In this case the client must abort or refetch the latest revision of foo and
retry.

### fooSessionId

The sessionId of the session that created the foo object.

## Setting the database engine for models

The database engine can be set either globally or on an individual model.

If the engine is specified on a model this will override any global value.

The database engine will not be changed or checked after the initial sync.

### Setting the database engine from the ENV

    DEFAULT_ENGINE=TokuDB node app.js

### Setting the database engine globally

    ImmutableCoreModel.defaultEngine('TokuDB')

When setting the defaultEngine the ImmutableCoreModel class object is returned
so that global configuration methods can be chained.

### Getting the global default database engine

    var defaultEngine = ImmutableCoreModel.defaultEngine()

### Setting the database engine on a model

    new ImmutableCoreModel({engine: 'TokuDB'})

## Setting the charset for models

The charset can be set either globally or on an individual model.

If the charset is specified on a model this will override any global value.

The charset will not be changed or checked after the initial sync.

### Setting the charset from the ENV

    DEFAULT_CHARSET=latin1 node app.js

### Setting the charset globally

    ImmutableCoreModel.defaultCharset('latin1')

When setting the defaultCharset the ImmutableCoreModel class object is returned
so that global configuration methods can be chained.

### Getting the global default charset

    var defaultCharset = ImmutableCoreModel.defaultCharset()

### Setting the charset on a model

    new ImmutableCoreModel({charset: 'latin1'})

## Creating a model with a JSON schema

    var fooModel = new ImmutableCoreModel({
        additionalProperties: false,
        errorMessage: {
            required: {
                foo: 'please enter foo'
            },
        },
        name: 'foo',
        properties: {
            foo: {
                type: 'string',
                default: 'foo',
            },
        },
        required: 'foo'
    })

The properties object is a list of [JSON Schema](https://spacetelescope.github.io/understanding-json-schema/index.html)
properties for the record data.

The schema specified here will be built into a schema for the complete object
including meta data columns (id, createTime, etc).

Properties with defaults will be added to the data.

Data type coercion will be performed so that numbers are converted to strings,
single element arrays are converted to scalar values, scalars are converted to
single element arrays, etc.

When additionalProperties:false is set any properties not in the schema will be
removed without throwing an error.

Immutable Core Model uses [ajv](https://www.npmjs.com/package/ajv) with the
following options to perform JSON schema validation:

    allErrors: true
    coerceTypes: 'array'
    removeAdditional: true
    useDefaults: true
    v5: true

[ajv-errors](https://www.npmjs.com/package/ajv-errors) is used to provide custom
error messages.

## Disabling schema validation at the model level

    var fooModel = new ImmutableCoreModel({
        name: 'foo',
        validate: false,
    })

JSON schema validation is enabled by default. To disable JSON schema validation
set the validate:false flag.

JSON schema validation can also be enabled and disabled for individual creat
and update calls.

Even if validation is disabled the schema and metaSchema properties of the model
will still be created.

## Disabling schema validation on create

    fooModel.createMeta({
        data: {foo: 'bar'},
        validate: false,
    })

## Disabling schema validation on update

    foo.updateMeta({
        data: {foo: 'bar'},
        validate: false,
    })

## Creating a model with queryable columns

    var fooModel = new ImmutableCoreModel({
        columns: {
            bam: 'boolean',
            bar: 'number',
            foo: 'string',
        },
        name: 'foo',
    })

While all of the object data for foo is stored in fooData it may be
necessary to query tables using column values.

Immutable Core Model allows for specifying additional columns that will be
created on the table schema.

Data for these columns will always match the values in fooData.

### Supported data types

Immutable Core Model supports the folowing data typesfor columns: boolean,
data, id, int, number, smallint, string and time.

boolean values must be either true or false.

data values must be JSON encodable and be able to be stored compressed in a
MEDIUMBLOB (~16MB).

id values must be specified as 32 char HEX string which will be stored as a
16 byte binary value.

strings have a maximum length of 255 characters. If strings exceed this length
then only the first 255 characters of the string will be stored in the
database column but the entire string will remain visible to the application.

numbers allow for a maximum of 27 digits before the decimal point and a
maximum of 9 digits after the decimal point.

time must be string that can be interpreted by MySQL as DATETIME.

### MySQL data type equivalents

Immutable Type | MySQL Type     |
---------------|----------------|
boolean        | TINYINT(1)     |
data           | MEDIUMBLOB     |
date           | DATE           |
id             | BINARY(16)     |
int            | BIGINT(20)     |
number         | DECIMAL(36,9)  |
smallint       | SMALLINT(5)    |
string         | VARCHAR(255)   |
time           | DATETIME(6)    |

### Creating a queryable column with non-default options

    var fooModel = new ImmutableCoreModel({
        columns: {
            foo: {
                default: 0,
                index: false,
                null: false,
                path: 'foo.bar',
                type: 'number',
            },
        },
        name: 'foo',
    })

When a column is created with a string value that string is used as the column
type and all other configuration options are set to defaults.

When a columen is created with an object value then non-default configuration
options can be set.

### Create a column with a unique index

    var accountModel = new ImmutableCoreModel({
        columns: {
            email: {
                type: 'string',
                unique: true,
            },
        },
        name: 'account',
    })

Because all the revisions of a record are stored in the same table a unique index
on a column would not allow revisions of a record to be inserted if the indexed
value did not change.

In order for this to work the value for the unique column is only inserted for
the first revision of the record and this column is left NULL for future
revisions unless the value changes.

To disable this behavior use the firstOnly:false option.

The firstOnly:true option is incompatible with the null:false option.

### Column options

Option Name | Default   | Description                                         |
------------|-----------|-----------------------------------------------------|
default     | null      | default value for column                            |
firstOnly   | true      | only apply unique index to original record          |
index       | true      | create index for column                             |
immutable   | false     | value cannot be changed after being set             |
null        | true      | allow null values                                   |
path        | undefined | path to get value from object (uses lodash _.get)   |
primary     | false     | column is primary key                               |
type        | undefined | data type (boolean|number|string)                   |
unique      | false     | create index as unique                              |
unsigned    | false     | create integer columns as unsigned                  |

## Creating multi-column indexes

    var fooModel = new ImmutableCoreModel({
        columns: {
            bam: 'boolean',
            bar: 'number',
        },
        indexes: [
            {
                columns: ['bam', 'bar'],
                unique: true
            },
        ],
        name: 'foo',
    })

Multi-column indexes must be specified separately from column specifications.

The indexes option can only be used for multi-column indexes and attempts to
create single column indexes will result in an error.

The unique flag controls whether or not the index is unique.

## Creating a model with unique id based on data only

    var fooModel = new ImmutableCoreModel({
        columns: {
            accountId: false,
            originalId: false,
            parentId: false,
        }
        idDataOnly: true,
        name: 'foo',
    })

When the idDataOnly flag is set then only the data values for an object will be
used to calculate the id for the object.

When creating models where the id is based only on the data it will usually not
make sense to allow revisions or ownership of records.

## Creating models with relations

    var fooModel = new ImmutableCoreModel({
        name: 'foo',
        relations: {
            bar: {},
        },
    })

    var barModel = new ImmutableCoreModel({
        columns: {
            fooId: {
                index: true,
                type: 'id',
            },
        },
        name: 'bar',
    })

Relations allow linked models to be created and queried from record objects.

In this example foo is related to bar.

The bar model must have either a fooId or fooOriginalId column.

For cases where the relation should apply to all revisions of a record the
originalId should be used. For cases where the relation only applies to a
specific revision the id should be used.

Relations are resolved at runtime which allows models to be defined in any
order.

### Creating models with relations linked via an intermediary table

    var fooModel = new ImmutableCoreModel({
        name: 'foo',
        relations: {
            bar: {via: 'bam'},
        },
    })

    var bamModel = new ImmutableCoreModel({
        columns: {
            barId: {
                index: true,
                null: false,
                type: 'id',
            },
            fooId: {
                index: true,
                null: false,
                type: 'id',
            },
            data: false,
            originalId: false,
            parentId: false,
        },
        name: 'bam',
    })

    var barModel = new ImmutableCoreModel({
        name: 'bar',
    })

The via option can be used to link two models via a third model.

The linking model should have either an id or originalId column from each of
the two models being linked.

### Creating a new related model

    var foo = fooModel.create(...)

    foo.create('bar', {foo: 'foo'})

To create a new related record the create method is call on an existing record
with the name of the related model and the data for the related record.

If the related record is linked via a third model the linking record will be
also be created.

### Creating a new related model with meta options

    foo.createMeta({
        data: {foo: 'foo'},
        relation: 'bar',
        session: session,
    })

The createMeta method can be used to set meta options for the create method.

### Selecting related records

    foo.select('bar')

The select method takes the name of a relation and queries all records related
to the record object.

The select method always returns a results object which must be used to fetch
or iterate over the related records.

### Querying related records

    foo.query({
        order: ['createTime'],
        relation: 'bar',
    })

The query method allows setting any of the options that are available with a
normal model query.

### Loading related records with query

    foo.query({
        where: {id: fooId},
        with: {
            bar: {
                order: ['createTime'],
            },
        },
    })

When doing a query for a single record by id related records can be loaded by
specifying the with option.

All records will be loaded so caution must be taken to make sure that this does
not become a performance and memory usage issue.

## Creating a model with transform functions

    var fooModel = new ImmutableCoreModel({
        name: 'foo',
        transform: {
            bam: (value, model, args) => {
                return 'bar'
            }
        }
    })

    fooModel.createMeta({
        data: {bam: 'foo'}
    })

Transform functions are called when creating or updating a record to modify
values set in the record data.

Transform functions will only be called if the value is defined in the data
passed to the create or update call.


CHANGELOG.md
# v6.0.0

Update dependencies to address vulns. Switch search backend from
elasticsearch to opensearch.

# v5.1.0

Add [ajv-errors](https://www.npmjs.com/package/ajv-errors) to replace
custom error message functionality that was previously a part of ajv.

# v5.0.0

Upgrade supporting packages to support Node 14. Fix breaking changes
caused by package upgrades.

## Major Version Upgrades

* ajv 8
* mysql2 2
* elasticsearch 16 (dev)
* redis 3 (dev)

# v4.0.0

Version 4.0.0 of Immutable Core Model switches to using
[mysql2](https://www.npmjs.com/package/mysql2) instead of
[immutable-database-mariasql](https://www.npmjs.com/package/immutable-database-mariasql)
as a database driver. This change requires updating many modules that depend
on Immutable Core Model as well as upating Immutable App app.js files.

mysql2 has a number of differences from mariasql that may break code
that interacts with raw database response data including:

* c and d columns are returned as integers instead of strings
* response from create/update commands is object instead of array.info

# v3.3.0

[Immutable Core](https://www.npmjs.com/package/immutable-core) introduced
breaking changes that require Immutable Core Model v3.3.0 or greater.

As of Immutable Core Model v3.3.0 the `accessControl` property for
ImmutableModel is deprecated and the global ImmutableAccessControl instance
will always be used.

# v3.0.0

Version 3.0.0 of Immutable Core Model is a major release with significant
new features and a few major breaking changes.

## Breaking changes

* support for actions eliminated
* support for `select` of custom columns eliminated
* support for using column names in where queries (e.g. fooId) eliminated
* changed class name ImmutableCoreModelInstance to ImmutableCoreModelRecord
* changed record unDelete method name to undelete

## New features

* caching of records, queries, and views
* resolve option to load related records referenced in record data
* n column auto increment primary key to track inserts and order
* select interface exposes then function that executes query
* c column to indicate if record is compressed
* d column to indicate if record is deleted
* int (64bit) and smallint (16bit) types


lib/sql.js
'use strict'

/* native modules */
const assert = require('assert')

/* npm modules */
const _ = require('lodash')

/* application modules */
const SqlSelect = require('./sql/select')

/* global variables */

// allowed column types and SQL definitions
const columnTypes = {
    boolean: 'TINYINT(1)',
    data: 'MEDIUMBLOB',
    date: 'DATE',
    id: 'BINARY(16)',
    int: 'BIGINT(20)',
    number: 'DECIMAL(36,9)',
    smallint: 'SMALLINT(5)',
    string: 'VARCHAR(255)',
    time: 'DATETIME(6)',
}
// invert column types and convert to lower case for mapping back mysql
// schema data to internal types
const columnTypesInverse = _.invert(
    _.mapValues(columnTypes, v => v.toLowerCase())
)

/* exports */
module.exports = {
    columnTypes: columnTypes,
    columnTypesInverse: columnTypesInverse,
    convertPrimaryToUnique: convertPrimaryToUnique,
    createColumn: createColumn,
    createIndex: createIndex,
    createTable: createTable,
    describe: describe,
    insert: insert,
    maxRecordNumber: maxRecordNumber,
    quoteName: quoteName,
    select: select,
    selectDeleted: selectDeleted,
    showCreateTable: showCreateTable,
    showIndexes: showIndexes,
    toBooleanInt: toBooleanInt,
    updateExtraColumns: updateExtraColumns,
}

/**
 * @function convertPrimaryToUnique
 *
 * generate sql to convert primary key to unique index
 *
 * @param {ImmutableCoreModel} model
 * @param {object} column
 *
 * @returns {string}
 *
 * @throws {Error}
 */
function convertPrimaryToUnique (model, column) {
    // build query
    var sql = `ALTER TABLE ${quoteName(model.name)} DROP PRIMARY KEY, ADD UNIQUE KEY (${quoteName(column.name)})`
    // reture sql string
    return sql
}

/**
 * @function createColumn
 *
 * generate sql for ALTER TABLE ADD COLUMN statement
 *
 * @param {ImmutableCoreModel} model
 * @param {object} column
 *
 * @returns {string}
 *
 * @throws {Error}
 */
function createColumn (model, column) {
    // build query
    var sql = `ALTER TABLE ${quoteName(model.name)} ADD COLUMN ${createTableColumn(model, column.name, column.spec)}`
    // add column first
    if (column.spec.first) {
        sql += ' FIRST'
    }
    // add column after
    if (column.spec.after) {
        sql += ` AFTER ${quoteName(column.spec.after)}`
    }
    // column has unique index
    if (column.spec.unique) {
        sql += `, ADD ${createTableUniqueKey(column.name, column.spec)}`
    }
    // column has non-unique index
    else if (column.spec.index) {
        sql += `, ADD ${createTableKey(column.name, column.spec)}`
    }
    // reture sql string
    return sql
}

/**
 * @function createIndex
 *
 * generate sql for ALTER TABLE ADD INDEX statement
 *
 * @param {ImmutableCoreModel} model
 * @param {object} column
 *
 * @returns {string}
 *
 * @throws {Error}
 */
function createIndex (model, column) {
    // build query
    var sql = 'ALTER TABLE '+quoteName(model.name)
    // column has unique index
    if (column.spec.unique) {
        sql = sql+' ADD '+createTableUniqueKey(column.name, column.spec)
    }
    // column has non-unique index
    else if (column.spec.index) {
        sql = sql+' ADD '+createTableKey(column.name, column.spec)
    }
    // reture sql string
    return sql
}

/**
 * @function createTable
 *
 * generate sql for CREATE TABLE statement
 *
 * @param {ImmutableCoreModel} model
 *
 * @returns {string}
 *
 * @throws {Error}
 */
function createTable (model) {
    var columnSql = []
    var keySql = []
    var primaryKeySql = []
    var uniqueKeySql = []
    // create sql for columns
    _.each(model.columnNames, name => {
        // get spec
        var spec = model.columns[name]
        // create column sql
        columnSql.push(createTableColumn(model, name, spec))
        // column has index
        if (spec.index || spec.primary || spec.unique) {
            // index is primary
            if (spec.primary) {
                primaryKeySql.push(createTablePrimaryKey(name, spec))
            }
            // index is unique
            else if (spec.unique) {
                uniqueKeySql.push(createTableUniqueKey(name, spec))
            }
            // index is not unique
            else {
                keySql.push(createTableKey(name, spec))
            }
        }
    })
    // require table to have at most one primary key
    assert.ok(primaryKeySql.length <= 1, model.name+': table must not have more than one primary key')
    // add any multi-column indexes
    _.each(model.indexes, (spec) => {
        // index is unique
        if (spec.unique) {
            uniqueKeySql.push(createTableUniqueKeyMulti(spec))
        }
        // index is not unique
        else {
            keySql.push(createTableKeyMulti(spec))
        }
    })
    // return create table sql
    return 'CREATE TABLE '+quoteName(model.name)+' ('
        + columnSql.concat(primaryKeySql, uniqueKeySql, keySql).join(', ')
        + ') ENGINE='+model.engine+' DEFAULT CHARSET='+model.charset
}

/**
 * @function describe
 *
 * generate sql for DESCRIBE statement
 *
 * @param {ImmutableCoreModel} model
 *
 * @returns {string}
 *
 * @throws {Error}
 */
function describe (model) {
    return `DESCRIBE ${quoteName(model.name)}`
}

/**
 * @function insert
 *
 * generate sql for INSERT statement
 *
 * @param {ImmutableCoreModel} model
 *
 * @returns {string}
 *
 * @throws {Error}
 */
function insert (model) {
    var insert = model.insertDelayed && model.engine === 'MyISAM' ? 'INSERT DELAYED' : 'INSERT'
    var columns = _.map(model.insertColumnNames, quoteName).join(', ')
    var values = _.map(model.insertColumnNames, columnName => placeholder(model, columnName, model.columns[columnName])).join(', ')
    return `${insert} INTO ${quoteName(model.name)} (${columns}) VALUES(${values})`
}

/**
 * @function maxRecordNumber
 *
 * generate sql for SELECT MAX(n) n FROM `tableName`
 *
 * @param {ImmutableCoreModel} model
 * @param {object} args
 *
 * @returns {string}
 */
function maxRecordNumber (model, args) {
    var sql = `SELECT MAX(n) n FROM ${quoteName(model.name)}`

    return {
        sql: sql,
    }
}

/**
 * @function placeholder
 *
 * create placeholder for value based on column type
 *
 * @params {object} model
 * @params {string} name
 * @params {object} spec
 *
 * @returns {string}
 *
 * @throws {Error}
 */
function placeholder (model, name, spec) {
    // if colum is id then needs to be wrapped in UNHEX
    if (spec.type === 'id' && model.binaryIds) {
        return `UNHEX(:${name})`
    }
    // if column is data and compression is in use it is hex
    else if (spec.type === 'data' && model.compression) {
        return `FROM_BASE64(:${name})`
    }
    else {
        return `:${name}`
    }
}

/**
 * @function quoteName
 *
 * quote name for table/column/index
 *
 * @params {string} name
 *
 * @returns {string}
 *
 * @throws {Error}
 */
function quoteName (name) {
    // require name to be string
    assert.equal(typeof name, 'string', 'name must be string')
    // do not allow quote symbol in name - all other name validation is
    // left to the database
    assert.ok(!name.match(/`/), 'invalid name')
    // return quoted name
    return '`'+name+'`'
}

/**
 * @function select
 *
 * generate sql for SELECT statement
 *
 * @param {ImmutableCoreModel} model
 * @param {object} args
 *
 * @returns {string}
 *
 * @throws {Error}
 */
function select (model, args) {
    return new SqlSelect(model, args)
}

/**
 * @function selectDeleted
 *
 * generate sql to select deleted record ids
 *
 * @param {ImmutableCoreModel} model
 *
 * @returns {string}
 */
function selectDeleted (model) {
    // get id column
    var idColumn = quoteName(model.columnName('id'))
    // wrap column in HEX if binary ids enabled
    idColumn = model.binaryIds ? `HEX(${idColumn})` : idColumn
    // build sql
    return `SELECT LOWER(${idColumn}) AS id FROM ${model.name}Delete`
}

/**
 * @function showCreateTable
 *
 * generate sql for SHOW CREATE TABLE statement
 *
 * @param {ImmutableCoreModel} model
 *
 * @returns {string}
 *
 * @throws {Error}
 */
function showCreateTable (model) {
    return 'SHOW CREATE TABLE '+quoteName(model.name)
}

/**
 * @function showIndexes
 *
 * generate sql for SHOW INDEX statement
 *
 * @param {ImmutableCoreModel} model
 *
 * @returns {string}
 *
 * @throws {Error}
 */
function showIndexes (model) {
    return 'SHOW INDEX FROM '+quoteName(model.name)
}

/* private functions */

/**
 * @function createTableColumn
 *
 * generate column sql for CREATE TABLE
 *
 * @param {ImmutableCoreModel} model
 * @param {string} name
 * @param {object} spec
 *
 * @returns {string}
 *
 * @throws {Error}
 */
function createTableColumn (model, name, spec) {
    // require spec to be object
    assert.ok(typeof spec === 'object', `${name} spec must be object`)
    // require valid type
    assert.ok(columnTypes[spec.type], `invalid type ${spec.type} for ${name}`)
    // build sql string
    var columnType = spec.type === 'id'
        ? (model.binaryIds ? 'BINARY(16)' : 'CHAR(32)')
        : columnTypes[spec.type]
    var sql = `${quoteName(name)} ${columnType}`
    // add unsigned flag if set
    if (spec.unsigned && (spec.type === 'int' || spec.type === 'smallint')) {
        sql += ' UNSIGNED'
    }
    // add not null if null specifically set to false
    if (spec.null === false) {
        sql += ' NOT NULL'
    }
    // add auto increment
    if (spec.auto) {
        sql += ' AUTO_INCREMENT PRIMARY KEY'
    }
    // add default value if set
    if (spec.default !== undefined) {
        // use 0/1 for boolean
        if (spec.type === 'boolean') {
            sql += ` DEFAULT ${spec.default ? 1 : 0}`
        }
        // otherwise use quoteNamed value
        else {
            sql += ` DEFAULT '${spec.default}'`
        }
    }
    // return sql string
    return sql
}

/**
 * @function createTableKey
 *
 * generate index sql for CREATE TABLE
 *
 * @param {string} name
 * @param {object} spec
 *
 * @returns {string}
 *
 * @throws {Error}
 */
function createTableKey (name, spec) {
    // require spec to be object
    assert.equal(typeof spec, 'object', name+' spec must be object')
    // return key sql
    return 'KEY ('+quoteName(name)+')'
}

/**
 * @function createTableKeyMulti
 *
 * generate multi-column non-unique index sql for CREATE TABLE
 *
 * @param {object} spec
 *
 * @returns {string}
 *
 * @throws {Error}
 */
function createTableKeyMulti (spec) {
    // require spec to be object
    assert.equal(typeof spec, 'object', 'spec must be object')
    // require columns to be array
    assert.ok(Array.isArray(spec.columns), 'index columns must be array')
    // return key sql
    return 'KEY ('+_.map(spec.columns, quoteName).join(', ')+')'
}

/**
 * @function createTablePrimaryKey
 *
 * generate unique index sql for CREATE TABLE
 *
 * @param {string} name
 * @param {object} spec
 *
 * @returns {string}
 *
 * @throws {Error}
 */
function createTablePrimaryKey (name, spec) {
    // require spec to be object
    assert.equal(typeof spec, 'object', name+' spec must be object')
    // return key sql
    return 'PRIMARY KEY ('+quoteName(name)+')'
}

/**
 * @function createTableUniqueKey
 *
 * generate unique index sql for CREATE TABLE
 *
 * @param {string} name
 * @param {object} spec
 *
 * @returns {string}
 *
 * @throws {Error}
 */
function createTableUniqueKey (name, spec) {
    // require spec to be object
    assert.equal(typeof spec, 'object', name+' spec must be object')
    // return key sql
    return 'UNIQUE KEY ('+quoteName(name)+')'
}

/**
 * @function createTableUniqueKeyMulti
 *
 * generate multi-column unique index sql for CREATE TABLE
 *
 * @param {object} spec
 *
 * @returns {string}
 *
 * @throws {Error}
 */
function createTableUniqueKeyMulti (spec) {
    // require spec to be object
    assert.equal(typeof spec, 'object', 'spec must be object')
    // require columns to be array
    assert.ok(Array.isArray(spec.columns), 'index columns must be array')
    // return key sql
    return 'UNIQUE KEY ('+_.map(spec.columns, quoteName).join(', ')+')'
}

/**
 * @function toBooleanInt
 *
 * convert various values to a boolean integer (0|1)
 *
 * @param {any} val
 *
 * @returns {integer}
 */
function toBooleanInt (val) {
    var type = typeof val
    // boolean
    if (type === 'boolean') {
        return val ? '1' : '0'
    }
    // string
    else if (type === 'string') {
        // true and 1 are only true string values
        return val === 'true' || val === '1' ? '1' : '0'
    }
    // number
    else if (type === 'number') {
        // 1 or greater is true
        return val >= 1 ? '1' : '0'
    }
    // all other values false
    else {
        return '0'
    }

}

/**
 * @function updateExtraColumns
 *
 * create sql to UPDATE extra columns on table and build params object for query
 *
 * @param {object} model
 * @param {object} data
 *
 * @returns {object}
 *
 * @throws {Error}
 */
function updateExtraColumns (model, instance) {
    // columns to update
    var updateColumns = []
    // placeholder => value params for query
    var params = {}
    // add instance id to params
    params.id = instance.id
    // build list of columns to update and param values
    _.each(model.extraColumns, (spec, name) => {
        // convert boolean to 1/0
        if (spec.type === 'boolean') {
            params[name] = toBooleanInt(_.get(instance.data, spec.path))
        }
        // set other values directly
        else {
            params[name] = _.get(instance.data, spec.path)
        }
        // add update sql
        updateColumns.push(`${quoteName(name)} = ${placeholder(model, name, spec)}`)
    })
    // build sql string
    var sql = `UPDATE ${quoteName(model.name)} SET ${updateColumns.join(', ')} WHERE ${quoteName(model.columnName('id'))} = ${placeholder(model, 'id', {type: 'id'})}`
    // return sql and params to perform query
    return {
        params: params,
        sql: sql,
    }
}